core/indigo-core/reaction/src/reaction_multistep_detector.cpp

```diff
@@ -238,6 +227,430 @@ void ReactionMultistepDetector::createSummBlocks()
+            // we never merge molecules with different sections of the same zone
+            if (zone2.has_value())
+            {
+                // disable catalysts merging
+                if ((_zones[zone1.value().first].zone_type == ZoneType::EArrow && zone1.value().second > 1) ||
+                    (_zones[zone2.value().first].zone_type == ZoneType::EArrow && zone2.value().second > 1))
+                    return std::nullopt;
+
+                // if both molecules are on the same zone - check if they are mergeable
+                for (auto& cz : comm_zones)
+                {
+                    auto it_oz2 = other_zones2.find(cz.first);
+                    if (it_oz2 != other_zones2.end())
+                    {
+                        for (auto& section : cz.second)
+                        {
+                            if (!(_zones[cz.first].zone_type == ZoneType::EPathWay && section > 1) && it_oz2->second.count(section ^ 1))
+                                return std::nullopt;
+
+                            if (it_oz2->second.count(section))
+                                return zone1;
+                        }
+                    }
+                }
+
+                // different zone types are not mergeable
+                if (_zones[zone1.value().first].zone_type != _zones[zone2.value().first].zone_type && comm_zones.empty())
+                    return std::nullopt;
+            }
+
+            if (!current_zone.has_value())
+                current_zone = zone1;
+        }
+        else if (!current_zone.has_value())
+            return std::nullopt;
+
+        auto zidx = current_zone.value().first;
+        // if molecules have different zones - check if they are mergeable
+        auto& hull1 = _components[mol_idx1].second;
+        auto& hull2 = _components[mol_idx2].second;
+
+        const auto& coords = _zones[zidx].origin_coordinates;
+        // check if both molecules are on the same zone continuation
+        switch (_zones[zidx].zone_type)
+        {
+        case ZoneType::EPlus:
+            if ((doesVerticalLineIntersectPolygon(coords[0].x, hull1) && doesVerticalLineIntersectPolygon(coords[0].x, hull2)) ||
+                (doesHorizontalLineIntersectPolygon(coords[0].x, hull1) && doesHorizontalLineIntersectPolygon(coords[0].x, hull2)))
+                return zone1 ? zone1 : current_zone;
+            break;
+        case ZoneType::EArrow:
+            if ((doesRayIntersectPolygon(coords[0], coords[1], hull1) && doesRayIntersectPolygon(coords[0], coords[1], hull2)) ||
+                (doesRayIntersectPolygon(coords[1], coords[0], hull1) && doesRayIntersectPolygon(coords[1], coords[0], hull2)))
+                return zone1 ? zone1 : current_zone;
+            break;
+        case ZoneType::EPathWay: {
+            auto c_it = coords.begin();
+            const Vec2f& head = *c_it++;
+            const Vec2f& spine_beg = *c_it++;
+            const Vec2f& spine_end = *c_it++;
```

Comment:
Why you use ```spine_end```? AFAIK spine is vertical so X coordinates of  ```spine_end``` and ```spine_beg``` are same.

Reply:
Yes, the same. I could use only spine_beg, but then I need to skip spine_end and use c_it++ instead of 
const Vec2f& spine_end = *c_it++;
I think that explicit reading of spine_end looks better. Added comment about it.